src/poc/miner_poc_denylist.erl

```diff
@@ -0,0 +1,200 @@
+            false
+    end.
+
+-spec get_version() -> pos_integer().
+get_version() ->
+    gen_server:call(?MODULE, get_version).
+
+-spec get_binary() -> binary().
+get_binary() ->
+    BaseDir = application:get_env(blockchain, base_dir, "data"),
+    DenyFile = filename:join([BaseDir, "denylist", "latest"]),
+    file:read_file(DenyFile).
+
+%% gen_server callbacks
+
+init([Type, URL, Key]) ->
+    %% load up any existing xor we have on disk
+    BaseDir = application:get_env(blockchain, base_dir, "data"),
+    DenyFile = filename:join([BaseDir, "denylist", "latest"]),
+    ok = filelib:ensure_dir(DenyFile),
+    %% filter version will be a positive integer or 0 depending on if we could load one from disk
+    FilterVersion = case filelib:is_regular(DenyFile) of
+                        true ->
+                            case file:read_file(DenyFile) of
+                                {ok, <<Version:8/integer, SignatureLen:16/integer-unsigned-little, Signature:SignatureLen/binary, Rest/binary>>} when Version == 1 ->
+                                    %% check signature is still valid against our key
+                                    case catch libp2p_crypto:verify(Rest, Signature, libp2p_crypto:b58_to_pubkey(Key)) of
+                                        true ->
+                                            <<Serial:32/integer-unsigned-little, FilterBin/binary>> = Rest,
+                                            case xorf:from_bin({exor, 32}, FilterBin) of
+                                                {ok, Filter} ->
+                                                    ok = persistent_term:put(?MODULE, {{binary_fuse, 32}, Filter}),
+                                                    Serial;
+                                                {error, Reason} ->
+                                                    lager:notice("failed to deserialize denylist from disk: ~p", [Reason]),
+                                                    0
+                                            end;
+                                        _ ->
+                                            lager:notice("failed to verify signature on denylist on disk"),
+                                            0
+                                    end;
+                                _ ->
+                                    lager:notice("unrecognized or corrupt denylist on disk"),
+                                    0
+                            end;
+                        false ->
+                            0
+                    end,
+    {ok, schedule_check(#state{type=Type, url=URL, key=Key, version=FilterVersion}, 0)}.
+
+handle_info(check, #state{type=github_release, url=URL, key=Key, version=Version, etag=Etag}=State) ->
+    %% pull the release definition
+    case httpc:request(get, {URL, [{"user-agent", "https://github.com/helium/miner"}] ++ [ {"if-none-match", Etag} || Etag /= undefined] }, [], [{body_format, binary}]) of
```

Comment:
Does this Etag need quotes to work like on S3?

Reply:
apparently not